src/marshal.c

```diff
@@ -227,13 +237,98 @@ int format_packet(bot_t *bot, void *_packet_data, void *_packet_raw)
     }
 
     varlen = varint32_encode(packet_raw - save, varint, 5);
-    if(packet_raw - save + varlen > len)
+    if(packet_raw - save + varlen > len) {
         return -1; // TODO: compression
+    }
     memmove(save + varlen, save, packet_raw - save);
     memmove(save, varint, varlen);
     return packet_raw - save + varlen;
 }
 
+/* TODO: Add bounds checking on packet_raw */
+void *marshal_slot(void *_packet_raw, slot_t *slot_data)
+{
+    int16_t block_id = slot_data->block_id;
+
+    char *packet_raw = _packet_raw;
+    if (-1 == block_id) { // Empty slot is 0xffff
+        return push(packet_raw, block_id, sizeof(block_id));
+    } else {
+        int8_t count = slot_data->count;
+        int16_t damage = slot_data->damage;
+        nbt_node *tree = slot_data->tree;
+
+        /* Convert little to big endian; flip the bytes. */
+        block_id = htons(block_id);
+        damage = htons(damage);
+
+        /* copy block_id (2 bytes) */
+        memcpy(packet_raw, &block_id, sizeof(block_id));
+        packet_raw += sizeof(block_id);
+
+        /* copy count (1 byte) */
+        memcpy(packet_raw, &count, sizeof(count));
+        packet_raw += sizeof(count);
+
+        /* copy damage (2 bytes) */
+        memcpy(packet_raw, &damage, sizeof(damage));
+        packet_raw += sizeof(damage);
+
+        if (tree) {
+            /* convert nbt tree to binary nbt data */
+            struct buffer nbt_data = nbt_dump_binary(tree);
+            memcpy(packet_raw, nbt_data.data, nbt_data.len);
+            packet_raw += nbt_data.len;
+            free(nbt_data.data);
+        } else {
+            memset(packet_raw, 0, sizeof(int8_t));
+            packet_raw += sizeof(int8_t);
+        }
+    }
+    return packet_raw;
+}
+
+/* Returns the length of the slot data so decode_packet knows how much of
+ * the raw packet data is for slot
+ **/
+uint64_t demarshal_slot(void *_data, slot_t *slot_data, uint64_t space_left)
+{
+    char *data = _data;
+
+    /* Read 2 bytes (block_id) */
+    memcpy(&slot_data->block_id, data, sizeof(slot_data->block_id));
+    data += sizeof(slot_data->block_id);
+    slot_data->block_id = ntohs(slot_data->block_id);
+    if(slot_data->block_id == -1) { /* No slot data */
+        return sizeof(int16_t); /* Must return length as 2 */
+    }
+
+    /* Read 1 byte (count) */
+    slot_data->count = *data;
+    data += sizeof(slot_data->count);
+
+    /* Read 2 bytes (damage) */
+    memcpy(&slot_data->damage, data, sizeof(slot_data->damage));
+    data += sizeof(slot_data->damage);
+    slot_data->damage = ntohs(slot_data->damage);
+
+    uint64_t nbt_length;
+    if(*data) { /* If next byte is non-zero */
+        nbt_node *tree = nbt_parse(data, space_left);
+        if (!tree) {/* Error */
+            exit(1);
+        }
+        slot_data->tree = tree;
+        struct buffer nbt_bin = nbt_dump_binary(tree);
+        nbt_length = nbt_bin.len;
```

Comment:
Space_left is not the length of the nbt data. It is the number of free bytes left in packet_data for the nbt data. This is to ensure that nbt_parse won't overflow packet_data.